src/go/k8s/controllers/redpanda/redpanda_decommission_controller.go

```diff
@@ -0,0 +1,588 @@
+)
+
+type DecommissionReconciler struct {
+	client.Client
+	// Namespace is where the pvc's are we are watching for
+	Namespace    string
+	OperatorMode bool
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *DecommissionReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		// nolint:dupl // these are required event filters
+		For(&v1.StatefulSet{}).WithEventFilter(UpdateEventFilter).Complete(r)
+}
+
+func (r *DecommissionReconciler) Reconcile(c context.Context, req ctrl.Request) (ctrl.Result, error) {
+	ctx, done := context.WithCancel(c)
+	defer done()
+
+	start := time.Now()
+	log := ctrl.LoggerFrom(ctx).WithName("DecommissionReconciler.Reconcile")
+
+	sts := &v1.StatefulSet{}
+
+	if err := r.Client.Get(ctx, req.NamespacedName, sts); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Examine if the object is under deletion
+	if !sts.ObjectMeta.DeletionTimestamp.IsZero() {
+		log.Info(fmt.Sprintf("the statefulset %q is being deleted", req.NamespacedName))
+		return ctrl.Result{}, nil
+	}
+
+	// check if there exists a "decommission" annotation, if so process the decommission, otherwise
+	// check if required.
+	if sts.Annotations == nil {
+		sts.Annotations = make(map[string]string)
+	}
+	timestamp, decommissionStarted := sts.Annotations[RedpandaDecommissionLabelKey]
+
+	var err error
+	var result ctrl.Result
+
+	log.Info(fmt.Sprintf("is decomission started: %t", decommissionStarted))
+
+	// nolint:nestif // this is ok
+	if decommissionStarted {
+		timeStampParsed, errParsing := time.Parse(time.RFC3339, timestamp)
+		if errParsing != nil {
+			err = fmt.Errorf("could not parse timestamp %q: %w", timestamp, errParsing)
+			result = ctrl.Result{}
+		} else {
+			result, err = r.reconcileDecommission(ctx, timeStampParsed, sts)
+			if err != nil {
+				log.Info(fmt.Sprintf("found error trying to reconcile decommission, we will requeue: %s", err))
+			}
+		}
+	} else {
+		result, _, err = r.verifyIfNeedDecommission(ctx, sts)
+		if err != nil {
+			log.Info(fmt.Sprintf("found error trying to verify if we need to decommission: %s", err))
+		}
+	}
+
+	// Log reconciliation duration
+	durationMsg := fmt.Sprintf("reconciliation finished in %s", time.Since(start).String())
+	if result.RequeueAfter > 0 {
+		durationMsg = fmt.Sprintf("%s, next run in %s", durationMsg, result.RequeueAfter.String())
+	}
+	if err != nil {
+		durationMsg = fmt.Sprintf("found error, we will requeue, reconciliation attempt finished in %s", time.Since(start).String())
+	}
+
+	log.Info(durationMsg)
+
+	return result, err
+}
+
+// TODO use release name based out of sts name and get helmrelease information from that.
+// may need to pull helm library to accomplish this.
+
+func (r *DecommissionReconciler) verifyIfNeedDecommission(ctx context.Context, sts *v1.StatefulSet) (ctrl.Result, map[string]interface{}, error) {
+	log := ctrl.LoggerFrom(ctx).WithName("DecommissionReconciler.verifyIfNeedDecommission")
```

Comment:
NIT: I like how @joe started to pass log down the function call chain and the name of the logger started to act as call stack. Could you consider to pass `log` to this function and call WithName without creating logger from context?